Google.Api.Generator/Generation/MetadataGenerator.cs

```diff
@@ -0,0 +1,91 @@
+﻿// Copyright 2021 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using Google.Gapic.Metadata;
+using static Google.Gapic.Metadata.GapicMetadata.Types;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Google.Api.Generator.Generation
+{
+    internal static class MetadataGenerator
+    {
+        private const string GapicMetadataSchemaVersion = "1.0";
+        private const string GapicMetadataFileDescription = "This file maps proto services/RPCs to the corresponding library clients/methods";
+        private const string GapicMetadataLanguage = "csharp";
+
+        private const string TransportKeyGrpc = "grpc";
+
+        internal static string GenerateGapicMetadataJson(List<ServiceDetails> allServiceDetails)
+        {
+            var gapicMetadata = new GapicMetadata
+            {
+                Schema = GapicMetadataSchemaVersion,
+                Comment = GapicMetadataFileDescription,
+                Language = GapicMetadataLanguage,
+                ProtoPackage = allServiceDetails.First().ProtoPackage,
+                LibraryPackage = allServiceDetails.First().Namespace,
+                Services = {
+                allServiceDetails.ToDictionary(
+                    serviceDetails => serviceDetails.ServiceName,
+                    ServiceForTransportMetadata
+                )}
+            };
+
+            JObject obj = JsonConvert.DeserializeObject<JObject>(
+                gapicMetadata.ToString(),
+                new JsonSerializerSettings { DateParseHandling = DateParseHandling.None }
+            );
+
+            if (obj == null)

```

Comment:
I don't think this is required - if the parse failed, I'd expect an exception rather than a null result.

Reply:
https://github.com/JamesNK/Newtonsoft.Json/blob/60be32f46ea1bf291f0ff041c75ba28d834d3895/Src/Newtonsoft.Json/JsonConvert.cs#L785 specifies the return type as `T?` and https://github.com/JamesNK/Newtonsoft.Json/blob/60be32f46ea1bf291f0ff041c75ba28d834d3895/Src/Newtonsoft.Json/Serialization/JsonSerializerInternalReader.cs#L162 does return `null`. I can't check right now but it seems like it'll only throw for non-nullable types.

Reply:
It looks like that would only happen if the `JsonReader` had no tokens in it - which will never happen given that we're building it from `gapicMetadata.ToString()`, which will always return the JSON representation of an object.

Reply:
True enough, that would require for proto-generated-code of ToString() to change significantly. I'll remove the null check.